File: reverseList.cpp
//
//  reverseList.cpp
//  leetcode
//
//  Created by Kael on 15/5/7.
//  Copyright (c) 2015年 Kael. All rights reserved.
//

#include <stdio.h>

/**
 * Definition for singly-linked list.
 */
  struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* tmp1 = nullptr;
        ListNode* tmp2 = nullptr;
        if (head == NULL || head->next == NULL) {
            return head;
        }
        while(head->next != NULL) {
            if (tmp1 == NULL) {
                tmp1 = new ListNode(head->val);
            }
            tmp2 = new ListNode(head->next->val);
            tmp2->next = tmp1;
            head = head->next;
            tmp1 = tmp2;
        }
        return tmp2;
    }
};

File: go/reverse-list.go
package main

import "fmt"

func main() {
	fmt.Println(reverseSlice([]int{1, 2, 3, 4, 5}))
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	a := make([]int, 0)
	tmp := copy(head)
	for {
		if tmp != nil {
			a = append(a, tmp.Val)
		} else {
			break
		}
		tmp = tmp.Next
	}
	b := reverseSlice(a)
	i := 0
	res := head
	for {
		if head != nil {
			head.Value = b[i]
		} else {
			break
		}
		head = head.Next
		i++
	}
	return res
}

func reverseSlice(s []int) []int {
	for i := 0; i < len(s)/2; i++ {
		tmp := s[i]
		s[i] = s[len(s)-i-1]
		s[len(s)-i-1] = tmp
	}
	return s
}
